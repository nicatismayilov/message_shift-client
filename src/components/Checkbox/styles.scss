.checkbox {
  width: 100%;
  display: flex;
  justify-content: flex-start;
  align-items: center;
  margin-bottom: 30px;

  &__input {
    display: none;
  }

  &__label {
    width: 100%;

    user-select: none;

    display: flex;
    align-items: center;

    cursor: pointer;
    transition: all 100ms;
    position: relative;

    color: #2b2c43;
    font-size: 15px;

    font-weight: 600;

    display: flex;

    &-indicator {
      height: 17px !important;
      width: 17px !important;
      background-color: #fff;
      transition: all 100ms;
      border: 2px solid #e7e7e7;
      border-radius: 4px;
      margin-right: 7.5px;
      position: relative;

      display: flex;
      align-items: center;
      justify-content: center;

      &::before {
        content: "";
        height: 85%;
        width: 85%;

        background-color: #090963;
        border-radius: 2px;
        transform: scale(0.5);
        opacity: 0;
        transform-origin: center;
        transition: all 50ms;
      }
    }

    &:hover &-indicator {
      border-color: #c9c7c7;
    }
  }

  &__input:checked + &__label > &__label-indicator {
    border-color: #090963;

    &::before {
      transform: scale(1);
      opacity: 1;
    }
  }
}
