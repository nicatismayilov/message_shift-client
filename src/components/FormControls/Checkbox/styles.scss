.checkbox {
  width: 100%;
  display: flex;
  justify-content: flex-start;
  align-items: center;
  margin-bottom: 30px;

  &__input {
    display: none;
  }

  &__label {
    width: 100%;

    user-select: none;

    display: flex;
    align-items: center;

    cursor: pointer;
    transition: all 100ms;
    position: relative;

    color: #2b2c43;
    font-size: 15px;

    font-weight: 500;

    display: flex;

    &-indicator {
      height: 17px !important;
      width: 17px !important;
      background-color: #fff;
      transition: all 100ms;
      border: 2px solid #e7e7e7;
      border-radius: 50%;
      margin-right: 7.5px;
      position: relative;

      // display: flex;
      // align-items: center;
      // justify-content: center;

      &::before {
        content: "";
        height: 85%;
        width: 85%;
        position: absolute;
        top: 50%;
        left: 50%;

        background-color: #090963;
        border-radius: 50%;
        transform: translate(-55%, -45%);
        opacity: 0;
        transition: all 50ms;
      }
    }

    &:hover &-indicator {
      border-color: #c9c7c7;
    }
  }

  &__input:checked + &__label > &__label-indicator {
    border-color: #090963;

    &::before {
      opacity: 1;
    }
  }
}
