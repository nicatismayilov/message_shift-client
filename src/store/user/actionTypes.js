export const AUHTENTICATE_USER_START = "AUHTENTICATE_USER_START";
export const AUHTENTICATE_USER_SUCCESS = "AUHTENTICATE_USER_SUCCESS";
export const AUHTENTICATE_USER_FAILURE = "AUHTENTICATE_USER_FAILURE";

export const FETCH_USER_START = "FETCH_USER_START";
export const FETCH_USER_SUCCESS = "FETCH_USER_SUCCESS";
export const FETCH_USER_FAILURE = "FETCH_USER_FAILURE";

export const REGISTER_USER_START = "REGISTER_USER_START";
export const REGISTER_USER_SUCCESS = "REGISTER_USER_SUCCESS";
export const REGISTER_USER_FAILURE = "REGISTER_USER_FAILURE";

export const LOGOUT_USER_START = "LOGOUT_USER_START";
export const LOGOUT_USER_SUCCESS = "LOGOUT_USER_SUCCESS";
export const LOGOUT_USER_FAILURE = "LOGOUT_USER_FAILURE";

export const CLEAR_USER_ERROR = "CLEAR_USER_ERROR";

export const SET_STAY_SIGNED_IN = "SET_STAY_SIGNED_IN";
