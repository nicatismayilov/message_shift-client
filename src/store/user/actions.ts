import { ActionTypes, Action } from "./types";
import { UserInfo } from "types/User";

// actions for fetching user data
export const fetchUserStart = (): Action => ({
	type: ActionTypes.FETCH_USER_START,
});

export const fetchUserSuccess = (userData: UserInfo): Action => ({
	type: ActionTypes.FETCH_USER_SUCCESS,
	payload: userData,
});

export const fetchUserFailure = (err: string): Action => ({
	type: ActionTypes.FETCH_USER_FAILURE,
	payload: err,
});

// action for cleaning user state
export const clearUserState = (): Action => ({
	type: ActionTypes.CLEAR_USER_STATE,
});
