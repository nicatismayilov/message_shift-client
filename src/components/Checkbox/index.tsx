import React, { useRef } from "react";

import generateKey from "utils/generateKey";

import "./styles.scss";

interface Props {
	checkhed: boolean;
	name: string;
	label?: string;
	onChange: (checked: boolean) => void;
}

const CheckBox: React.FC<Props> = (props) => {
	const { checkhed = false, name, label } = props;
	const { onChange } = props;

	const id = useRef(generateKey(name));

	const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
		const { checked } = e.target;

		onChange(checked);
	};

	return (
		<div className='checkbox'>
			<input
				type='checkbox'
				name={name}
				id={id.current}
				checked={checkhed}
				className='checkbox__input'
				onChange={handleChange}
			/>
			<label htmlFor={id.current} className='checkbox__label'>
				<span className='checkbox__label-indicator'></span>
				{label}
			</label>
		</div>
	);
};

export default CheckBox;
