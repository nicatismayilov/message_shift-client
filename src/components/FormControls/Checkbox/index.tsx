import React, { useRef } from "react";

import generateKey from "utils/generateKey";

import "./styles.scss";

interface Props {
	checkhed: boolean;
	name: string;
	label?: string;
	onChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
}

const CheckBox: React.FC<Props> = (props) => {
	const { checkhed = false, name, label } = props;
	const { onChange } = props;

	const id = useRef(generateKey(name));

	return (
		<div className='checkbox'>
			<input
				type='checkbox'
				name={name}
				id={id.current}
				checked={checkhed}
				className='checkbox__input'
				onChange={onChange}
			/>
			<label htmlFor={id.current} className='checkbox__label'>
				<span className='checkbox__label-indicator'></span>
				{label}
			</label>
		</div>
	);
};

export default CheckBox;
