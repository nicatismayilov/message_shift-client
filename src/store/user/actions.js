import * as actionTypes from "./actionTypes";

// actions for user authentication
export const authenticateUserStart = (credentials) => ({
	type: actionTypes.AUHTENTICATE_USER_START,
	payload: credentials,
});

export const authenticateUserSuccess = (token) => ({
	type: actionTypes.AUHTENTICATE_USER_SUCCESS,
	payload: token,
});

export const authenticateUserFailure = (err) => ({
	type: actionTypes.AUHTENTICATE_USER_FAILURE,
	payload: err,
});

// actions for fetching user data
export const fetchUserStart = () => ({
	type: actionTypes.FETCH_USER_START,
});

export const fetchUserSuccess = (userData) => ({
	type: actionTypes.FETCH_USER_SUCCESS,
	payload: userData,
});

export const fetchUserFailure = (err) => ({
	type: actionTypes.FETCH_USER_FAILURE,
	payload: err,
});

// actions for user registration
export const registerUserStart = (userData) => ({
	type: actionTypes.REGISTER_USER_START,
	payload: userData,
});

export const registerUserSuccess = (token) => ({
	type: actionTypes.REGISTER_USER_SUCCESS,
	payload: token,
});

export const registerUserFailure = (err) => ({
	type: actionTypes.REGISTER_USER_FAILURE,
	payload: err,
});

// action for logout
export const logout = () => ({
	type: actionTypes.LOGOUT,
});
