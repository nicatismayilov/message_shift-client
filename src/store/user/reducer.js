import * as actionTypes from "./actionTypes";

const initialState = {
	isAuth: false,
	token: "",
	user: {},
	loading: false,
	error: "",
	staySignedIn: false,
};

const reducer = (state = initialState, action) => {
	switch (action.type) {
		case actionTypes.AUHTENTICATE_USER_START:
		case actionTypes.REGISTER_USER_START:
			return {
				...state,
				loading: true,
				isAuth: false,
			};

		case actionTypes.LOGOUT_USER_START:
			return {
				...state,
				loading: true,
			};

		case actionTypes.AUHTENTICATE_USER_SUCCESS:
		case actionTypes.REGISTER_USER_SUCCESS:
			return {
				...state,
				loading: false,
				isAuth: true,
				token: action.payload,
			};

		case actionTypes.AUHTENTICATE_USER_FAILURE:
		case actionTypes.REGISTER_USER_FAILURE:
			return {
				...state,
				error: action.payload,
				loading: false,
				isAuth: false,
				token: "",
			};

		case actionTypes.FETCH_USER_START:
			return {
				...state,
				loading: true,
			};

		case actionTypes.FETCH_USER_SUCCESS:
			return {
				...state,
				user: action.payload,
				loading: false,
			};

		case actionTypes.LOGOUT_USER_SUCCESS:
			return initialState;

		case actionTypes.FETCH_USER_FAILURE:
		case actionTypes.LOGOUT_USER_FAILURE:
			return {
				...state,
				error: action.payload,
				loading: false,
			};

		case actionTypes.CLEAR_USER_ERROR:
			return {
				...state,
				error: "",
			};

		case actionTypes.SET_STAY_SIGNED_IN:
			return {
				...state,
				staySignedIn: action.payload,
			};

		default:
			return state;
	}
};

export default reducer;
