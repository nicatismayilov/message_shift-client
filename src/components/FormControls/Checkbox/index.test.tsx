import React from "react";
import { render } from "@testing-library/react";
import user from "@testing-library/user-event";

import Checkbox from "./index";

const handleChange = jest.fn();
const handleBlur = jest.fn();

describe("Checkbox component", () => {
	it("should render properly", () => {
		const { baseElement } = render(
			<Checkbox checkhed={false} name='checkbox' onChange={handleChange} onBlur={handleBlur} />
		);

		expect(baseElement).toMatchSnapshot();
	});

	it("should have correct name and checked value", () => {
		const mockChecked = true;
		const mockName = "Test Checkbox";

		const { getByTestId } = render(
			<Checkbox
				checkhed={mockChecked}
				name={mockName}
				onChange={handleChange}
				onBlur={handleBlur}
			/>
		);

		const CheckboxInput = getByTestId("checkbox-input");

		expect(CheckboxInput).toHaveProperty("checked", mockChecked);
		expect(CheckboxInput).toHaveProperty("name", mockName);
	});

	it("should call handleChange according to user interaction", () => {
		const mockChecked = true;
		const mockName = "Test Checkbox";

		const { getByTestId } = render(
			<Checkbox
				checkhed={mockChecked}
				name={mockName}
				onChange={handleChange}
				onBlur={handleBlur}
			/>
		);

		const CheckboxInput = getByTestId("checkbox-input");

		expect(CheckboxInput).toHaveProperty("checked", mockChecked);
		expect(CheckboxInput).toHaveProperty("name", mockName);
		user.click(CheckboxInput);
		user.click(CheckboxInput);

		expect(handleChange).toHaveBeenCalled();
		expect(handleChange).toHaveBeenCalledTimes(2);
	});
});
