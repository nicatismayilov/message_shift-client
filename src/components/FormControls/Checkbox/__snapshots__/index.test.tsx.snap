// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox component should render properly 1`] = `
<body>
  <div>
    <div
      class="checkbox"
    >
      <input
        class="checkbox__input"
        data-testid="checkbox-input"
        id="17638775e30-checkbox-3.be17c8a0520ac"
        name="checkbox"
        type="checkbox"
      />
      <label
        class="checkbox__label"
        for="17638775e30-checkbox-3.be17c8a0520ac"
      >
        <span
          class="checkbox__label-indicator"
        />
      </label>
    </div>
  </div>
</body>
`;
